File: research2/python/single_thread.py
import os
import sys
import time
from pathlib import Path
from PIL import Image

INPUT_DIR  = Path("research2/images")
OUTPUT_DIR = Path("research2/output/python_single")

EXTENSIONS = {".jpg", ".jpeg", ".png"}
TARGET_SIZE = (64, 64)


def process_image(src: Path, dst: Path) -> bool:
    try:
        with Image.open(src) as img:
            result = img.convert("L").resize(TARGET_SIZE, Image.NEAREST)
            result.save(dst)
        return True
    except Exception:
        return False


def main() -> None:
    if not INPUT_DIR.is_dir():
        print(f"Error: input directory '{INPUT_DIR}' not found.")
        sys.exit(1)

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    files = [p for p in INPUT_DIR.iterdir() if p.suffix.lower() in EXTENSIONS]

    start = time.perf_counter()
    ok = 0

    for src in files:
        dst = OUTPUT_DIR / src.name
        if process_image(src, dst):
            ok += 1

    elapsed = time.perf_counter() - start
    print(f"Time: {elapsed:.4f}s | Processed: {ok}/{len(files)}")


if __name__ == "__main__":
    main()

File: research2/rust/src/bin/single_thread.rs
use image::imageops::FilterType;
use image::ImageError;
use std::path::{Path, PathBuf};
use std::time::Instant;
use std::{fs, process};

const INPUT_DIR: &str = "research2/images";
const OUTPUT_DIR: &str = "research2/output/rust_single";
const TARGET_W: u32 = 64;
const TARGET_H: u32 = 64;

fn process_image(src: &Path, dst: &Path) -> Result<(), ImageError> {
    let img = image::open(src)?;
    let gray = img.to_luma8();
    let resized = image::imageops::resize(&gray, TARGET_W, TARGET_H, FilterType::Nearest);
    resized.save(dst)?;
    Ok(())
}

fn main() {
    let input_dir = Path::new(INPUT_DIR);
    let output_dir = Path::new(OUTPUT_DIR);

    if !input_dir.is_dir() {
        eprintln!("Error: input directory '{}' not found.", INPUT_DIR);
        process::exit(1);
    }

    fs::create_dir_all(output_dir).expect("Failed to create output directory");

    let files: Vec<PathBuf> = fs::read_dir(input_dir)
        .expect("Failed to read input directory")
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| {
            matches!(
                p.extension().and_then(|e| e.to_str()),
                Some("jpg") | Some("jpeg") | Some("png")
            )
        })
        .collect();

    let total = files.len();
    let start = Instant::now();
    let mut ok: usize = 0;

    for src in &files {
        let dst = output_dir.join(src.file_name().unwrap());
        if process_image(src, &dst).is_ok() {
            ok += 1;
        }
    }

    let elapsed = start.elapsed().as_secs_f64();
    println!("Time: {:.4}s | Processed: {}/{}", elapsed, ok, total);
}
